File: tldw_Server_API/app/api/v1/schemas/study_suggestions.py
"""Shared Pydantic schemas for study-suggestion status and action contracts."""

from typing import Any, Literal

from pydantic import BaseModel, Field


SuggestionApiStatus = Literal["none", "pending", "ready", "failed"]


class SuggestionRefreshRequest(BaseModel):
    """Request body for explicitly refreshing study suggestions."""

    reason: str | None = None


class SuggestionStatusResponse(BaseModel):
    """Status payload returned when checking suggestion state for an anchor."""

    anchor_type: str
    anchor_id: int
    status: SuggestionApiStatus
    job_id: int | None = None
    snapshot_id: int | None = None


class SuggestionActionResponse(BaseModel):
    """Response describing the concrete study artifact opened or generated."""

    disposition: Literal["opened_existing", "generated"]
    snapshot_id: int
    selection_fingerprint: str
    target_service: Literal["quiz", "flashcards"]
    target_type: str
    target_id: str


class SuggestionJobSummary(BaseModel):
    """Serialized Jobs metadata returned by study-suggestion endpoints."""

    id: int
    status: str


class SuggestionJobAcceptedResponse(BaseModel):
    """Accepted response for queued study-suggestion refresh work."""

    job: SuggestionJobSummary


class SuggestionSnapshotResource(BaseModel):
    """Frozen snapshot payload returned to clients."""

    id: int
    service: str
    activity_type: str
    anchor_type: str
    anchor_id: int
    suggestion_type: str
    status: str
    payload: dict[str, Any] | list[Any] | str
    user_selection: dict[str, Any] | list[Any] | str | None = None
    refreshed_from_snapshot_id: int | None = None
    created_at: str | None = None
    last_modified: str | None = None


class SuggestionSnapshotResponse(BaseModel):
    """Read response for a snapshot plus live evidence hydration."""

    snapshot: SuggestionSnapshotResource
    live_evidence: dict[str, Any]


class SuggestionActionRequest(BaseModel):
    """Request body for triggering a study-suggestion follow-up action."""

    target_service: Literal["quiz", "flashcards"]
    target_type: str = Field(..., min_length=1)
    action_kind: str = Field(..., min_length=1)
    selected_topic_ids: list[str] = Field(default_factory=list)
    selected_topic_edits: list[dict[str, str]] = Field(default_factory=list)
    manual_topic_labels: list[str] = Field(default_factory=list)
    has_explicit_selection: bool = False
    generator_version: str = Field(default="v1", min_length=1)
    force_regenerate: bool = False

File: apps/packages/ui/src/services/studySuggestions.ts
import { bgRequest } from "@/services/background-proxy"
import type { AllowedPath } from "@/services/tldw/openapi-guard"

export type SuggestionAnchorType = "quiz_attempt" | "flashcard_review_session"
export type SuggestionStatus = "none" | "pending" | "ready" | "failed"
export type SuggestionTargetService = "quiz" | "flashcards"
export type SuggestionDisposition = "opened_existing" | "generated"

export type StudySuggestionStatusResponse = {
  anchor_type: SuggestionAnchorType
  anchor_id: number
  status: SuggestionStatus
  job_id?: number | null
  snapshot_id?: number | null
}

export type StudySuggestionSnapshotTopic = {
  id: string
  display_label?: string | null
  topic_key?: string | null
  canonical_label?: string | null
  evidence_reasons?: string[] | null
  type?: string | null
  status?: string | null
  selected?: boolean | null
  source_type?: string | null
  source_id?: string | null
}

export type StudySuggestionSnapshotResource = {
  id: number
  service: string
  activity_type: string
  anchor_type: SuggestionAnchorType
  anchor_id: number
  suggestion_type: string
  status: string
  payload: {
    summary?: Record<string, unknown>
    counts?: Record<string, unknown>
    topics?: StudySuggestionSnapshotTopic[]
    [key: string]: unknown
  } | Record<string, unknown> | string | null
  user_selection?: Record<string, unknown> | unknown[] | string | null
  refreshed_from_snapshot_id?: number | null
  created_at?: string | null
  last_modified?: string | null
}

export type StudySuggestionSnapshotResponse = {
  snapshot: StudySuggestionSnapshotResource
  live_evidence: Record<string, unknown>
}

export type StudySuggestionRefreshRequest = {
  reason?: string | null
}

export type StudySuggestionRefreshResponse = {
  job: {
    id: number
    status: string
  }
}

export type StudySuggestionActionRequest = {
  targetService: SuggestionTargetService
  targetType: string
  actionKind: string
  selectedTopicIds: string[]
  selectedTopicEdits?: Array<{
    id: string
    label: string
  }>
  manualTopicLabels?: string[]
  hasExplicitSelection?: boolean
  generatorVersion?: string
  forceRegenerate?: boolean
}

export type StudySuggestionActionResponse = {
  disposition: SuggestionDisposition
  snapshot_id: number
  selection_fingerprint: string
  target_service: SuggestionTargetService
  target_type: string
  target_id: string
}

const encodeSegment = (value: string | number): string =>
  encodeURIComponent(String(value))

export const parseStudySuggestionTargetId = (
  targetId: string | null | undefined
): number | undefined => {
  if (typeof targetId !== "string") {
    return undefined
  }

  const normalized = targetId.trim()
  if (!normalized) {
    return undefined
  }

  if (/^\d+$/.test(normalized)) {
    const parsed = Number.parseInt(normalized, 10)
    return Number.isFinite(parsed) && parsed > 0 ? parsed : undefined
  }

  const prefixedMatch = /^(?:quiz|deck)-(\d+)$/i.exec(normalized)
  if (!prefixedMatch) {
    return undefined
  }

  const parsed = Number.parseInt(prefixedMatch[1], 10)
  return Number.isFinite(parsed) && parsed > 0 ? parsed : undefined
}

export async function getStudySuggestionAnchorStatus(
  anchorType: SuggestionAnchorType,
  anchorId: number,
  options?: { signal?: AbortSignal }
): Promise<StudySuggestionStatusResponse> {
  return await bgRequest<StudySuggestionStatusResponse, AllowedPath, "GET">({
    path: `/api/v1/study-suggestions/anchors/${encodeSegment(anchorType)}/${encodeSegment(anchorId)}/status` as any,
    method: "GET",
    abortSignal: options?.signal
  })
}

export async function getStudySuggestionSnapshot(
  snapshotId: number,
  options?: { signal?: AbortSignal }
): Promise<StudySuggestionSnapshotResponse> {
  return await bgRequest<StudySuggestionSnapshotResponse, AllowedPath, "GET">({
    path: `/api/v1/study-suggestions/snapshots/${encodeSegment(snapshotId)}` as any,
    method: "GET",
    abortSignal: options?.signal
  })
}

export async function refreshStudySuggestionSnapshot(
  snapshotId: number,
  request: StudySuggestionRefreshRequest = {},
  options?: { signal?: AbortSignal }
): Promise<StudySuggestionRefreshResponse> {
  return await bgRequest<StudySuggestionRefreshResponse, AllowedPath, "POST">({
    path: `/api/v1/study-suggestions/snapshots/${encodeSegment(snapshotId)}/refresh` as any,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: request,
    abortSignal: options?.signal
  })
}

export async function performStudySuggestionAction(
  snapshotId: number,
  request: StudySuggestionActionRequest,
  options?: { signal?: AbortSignal }
): Promise<StudySuggestionActionResponse> {
  return await bgRequest<StudySuggestionActionResponse, AllowedPath, "POST">({
    path: `/api/v1/study-suggestions/snapshots/${encodeSegment(snapshotId)}/actions` as any,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: {
      target_service: request.targetService,
      target_type: request.targetType,
      action_kind: request.actionKind,
      selected_topic_ids: request.selectedTopicIds,
      selected_topic_edits: request.selectedTopicEdits ?? [],
      manual_topic_labels: request.manualTopicLabels ?? [],
      has_explicit_selection: request.hasExplicitSelection ?? false,
      generator_version: request.generatorVersion,
      force_regenerate: request.forceRegenerate ?? false
    },
    abortSignal: options?.signal
  })
}
